File: integrations/sqlite/python/src/purra_sqlite/recovery_schedule.py
"""Persistent scheduling hints; never execution ownership or authorization."""
import time


def _integer(value):
    if type(value) is not int or not 0 <= value < 9007199254740991:
        raise ValueError('invalid recovery schedule integer')
    return value


class SqliteRecoverySchedule:
    def __init__(self, storage, *, interval_ms=1000, max_backoff_ms=30000, clock_ms=None, max_failures=None):
        if not 0 < _integer(interval_ms) <= _integer(max_backoff_ms) <= 2147483647:
            raise ValueError('invalid recovery schedule intervals')
        if max_failures is not None and not 1 <= _integer(max_failures) <= 31:
            raise ValueError("max_failures must be between 1 and 31")
        self._max_failures = max_failures
        self._storage = storage
        self._interval = interval_ms
        self._maximum = max_backoff_ms
        self._clock = clock_ms or (lambda: int(time.time() * 1000))

    @staticmethod
    def _row(extra, run_id):
        if not isinstance(run_id, str) or not run_id.strip():
            raise ValueError('run id required')
        rows = extra.get('recoverySchedule', {})
        if not isinstance(rows, dict): raise ValueError('invalid recovery schedule')
        row = rows.get(run_id, {'revision': _integer(extra.get('recoveryScheduleRevision', 0)), 'failures': 0, 'notBeforeMs': 0})
        if not isinstance(row, dict) or set(row) != {'revision', 'failures', 'notBeforeMs'}:
            raise ValueError('invalid recovery schedule')
        for value in row.values(): _integer(value)
        return row

    async def check(self, run_id):
        """Observe scheduling eligibility, not permission to execute."""
        async with self._storage._transaction(read_only=True, with_journal=False) as session:
            row = self._row(session.extra, run_id)
            reason = ("retry_exhausted" if self._max_failures is not None and row['failures'] >= self._max_failures
                      else "retry_not_due" if row['notBeforeMs'] > _integer(self._clock()) else None)
            return {"revision": row['revision'] if reason is None else None, "reason": reason}

    async def ready(self, run_id):
        return (await self.check(run_id))["revision"]

    async def wake(self, run_id):
        async with self._storage._transaction(with_journal=False) as session:
            return wake_recovery_schedule(session.extra, run_id)

    async def settle(self, run_id, revision, failed):
        _integer(revision)
        if type(failed) is not bool: raise ValueError('failed must be boolean')
        async with self._storage._transaction(with_journal=False) as session:
            row = self._row(session.extra, run_id)
            if row['revision'] != revision: return False
            failures = min(row['failures'] + 1, 31) if failed else 0
            delay = min(self._maximum, self._interval * 2 ** failures)
            session.extra.setdefault('recoverySchedule', {})[run_id] = {
                'revision': next_revision(session.extra, revision), 'failures': failures,
                'notBeforeMs': _integer(_integer(self._clock()) + delay),
            }
            return True


def wake_recovery_schedule(extra, run_id, *, existing_only=False):
    """Mutate the caller's transaction; replayed commands must not call this."""
    row = SqliteRecoverySchedule._row(extra, run_id)
    if existing_only and run_id not in extra.get('recoverySchedule', {}):
        return None
    revision = next_revision(extra, row['revision'])
    extra.setdefault('recoverySchedule', {})[run_id] = {
        'revision': revision, 'failures': 0, 'notBeforeMs': 0,
    }
    return revision


def next_revision(extra, current):
    revision = _integer(max(_integer(extra.get('recoveryScheduleRevision', 0)), current) + 1)
    extra['recoveryScheduleRevision'] = revision
    return revision


def remove_recovery_schedule(extra, run_id):
    row = SqliteRecoverySchedule._row(extra, run_id)
    if run_id not in extra.get('recoverySchedule', {}):
        return False
    next_revision(extra, row['revision'])
    del extra['recoverySchedule'][run_id]
    return True

File: integrations/sqlite/typescript/src/recovery-schedule.ts
/** Persistent scheduling hints; never ownership or authorization. */
type Extra = Record<string, any>;
type Transaction = <T>(operation: (extra: Extra) => Promise<T>, readOnly: boolean) => Promise<T>;
function integer(value: number): number {
  if (!Number.isSafeInteger(value) || value < 0 || value >= Number.MAX_SAFE_INTEGER) throw new TypeError('invalid recovery schedule integer');
  return value;
}
export class SqliteRecoverySchedule {
  constructor(private readonly transaction: Transaction,
    private readonly options: { intervalMs?: number; maxBackoffMs?: number; clockMs?: () => number } = {}) {
    this.options = Object.freeze({ ...options });
    const interval = integer(options.intervalMs ?? 1000), maximum = integer(options.maxBackoffMs ?? 30000);
    if (!(interval > 0 && interval <= maximum && maximum <= 2147483647)) throw new TypeError('invalid recovery schedule intervals');
  }
  static row(extra: Extra, runId: string): { revision: number; failures: number; notBeforeMs: number } {
    if (typeof runId !== 'string' || !runId.trim()) throw new TypeError('run id required');
    const rows = extra.recoverySchedule === undefined ? {} : extra.recoverySchedule;
    if (typeof rows !== 'object' || rows === null || Array.isArray(rows)) throw new TypeError('invalid recovery schedule');
    const row = Object.hasOwn(rows, runId) ? rows[runId] : { revision: integer(extra.recoveryScheduleRevision === undefined ? 0 : extra.recoveryScheduleRevision), failures: 0, notBeforeMs: 0 };
    if (row === null || typeof row !== 'object' || Object.keys(row).sort().join() !== 'failures,notBeforeMs,revision') throw new TypeError('invalid recovery schedule');
    for (const value of Object.values(row)) integer(value as number);
    return row;
  }
  static save(extra: Extra, runId: string, row: object): void {
    const rows = Object.assign(Object.create(null), extra.recoverySchedule ?? {});
    rows[runId] = row; extra.recoverySchedule = rows;
  }
  async ready(runId: string): Promise<number | null> {
    return this.transaction(async extra => {
      const row = SqliteRecoverySchedule.row(extra, runId);
      return row.notBeforeMs <= integer((this.options.clockMs ?? Date.now)()) ? row.revision : null;
    }, true);
  }
  async wake(runId: string): Promise<number> {
    return this.transaction(async extra => {
      return wakeRecoverySchedule(extra, runId)!;
    }, false);
  }
  async settle(runId: string, revision: number, failed: boolean): Promise<boolean> {
    integer(revision);
    if (typeof failed !== 'boolean') throw new TypeError('failed must be boolean');
    return this.transaction(async extra => {
      const row = SqliteRecoverySchedule.row(extra, runId);
      if (row.revision !== revision) return false;
      const failures = failed ? Math.min(row.failures + 1, 31) : 0;
      const delay = Math.min(this.options.maxBackoffMs ?? 30000, (this.options.intervalMs ?? 1000) * 2 ** failures);
      SqliteRecoverySchedule.save(extra, runId, { revision: nextRevision(extra, revision), failures,
        notBeforeMs: integer(integer((this.options.clockMs ?? Date.now)()) + delay) });
      return true;
    }, false);
  }
}

export function wakeRecoverySchedule(extra: Extra, runId: string, existingOnly = false): number | undefined {
  const row = SqliteRecoverySchedule.row(extra, runId);
  if (existingOnly && !Object.hasOwn(extra.recoverySchedule ?? {}, runId)) return undefined;
  const revision = nextRevision(extra, row.revision);
  SqliteRecoverySchedule.save(extra, runId, { revision, failures: 0, notBeforeMs: 0 });
  return revision;
}

function nextRevision(extra: Extra, current: number): number {
  const revision = integer(Math.max(integer(extra.recoveryScheduleRevision === undefined ? 0 : extra.recoveryScheduleRevision), current) + 1);
  extra.recoveryScheduleRevision = revision;
  return revision;
}

export function removeRecoverySchedule(extra: Extra, runId: string): boolean {
  const row = SqliteRecoverySchedule.row(extra, runId);
  if (!Object.hasOwn(extra.recoverySchedule ?? {}, runId)) return false;
  nextRevision(extra, row.revision);
  delete extra.recoverySchedule[runId];
  return true;
}
